<a href="https://colab.research.google.com/github/DangHuuLong/Ai-Recruiter-Mini-Ai-Service/blob/experiment%2Ffine-tune-similarity-colab-v0.2/notebooks/fine_tune_similarity_v0.2_colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fine-tune CV-JD Similarity Model v0.2

This notebook is a thin Colab wrapper around `training/fine_tune_similarity.py`.

Use GPU runtime before running the training cells.

## 1. Clone repository and checkout branch

In [25]:
import os

if not os.path.exists('/content/Ai-Recruiter-Mini-Ai-Service'):
    !git clone https://github.com/DangHuuLong/Ai-Recruiter-Mini-Ai-Service.git /content/Ai-Recruiter-Mini-Ai-Service

%cd /content/Ai-Recruiter-Mini-Ai-Service
!git checkout experiment/fine-tune-similarity-colab-v0.2
!git pull origin experiment/fine-tune-similarity-colab-v0.2

/content/Ai-Recruiter-Mini-Ai-Service
Already on 'experiment/fine-tune-similarity-colab-v0.2'
Your branch is up to date with 'origin/experiment/fine-tune-similarity-colab-v0.2'.
remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (29/29), done.
remote: Compressing objects: 100% (21/21), done.
remote: Total 24 (delta 12), reused 4 (delta 3), pack-reused 0 (from 0)
Unpacking objects: 100% (24/24), 11.39 KiB | 833.00 KiB/s, done.
From https://github.com/DangHuuLong/Ai-Recruiter-Mini-Ai-Service
 * branch            experiment/fine-tune-similarity-colab-v0.2 -> FETCH_HEAD
   907747a..2c83a86  experiment/fine-tune-similarity-colab-v0.2 -> origin/experiment/fine-tune-similarity-colab-v0.2
Updating 907747a..2c83a86
Fast-forward
 notebooks/fine_tune_similarity_v0.2_colab.ipynb | 871 ++++++++++++++++++++++--
 training/fine_tune_similarity.py                |  32 +-
 2 files changed, 846 insertions(+), 57 deletions(-)


In [14]:
%cd /content/Ai-Recruiter-Mini-Ai-Service
!git status

/content/Ai-Recruiter-Mini-Ai-Service
On branch experiment/fine-tune-similarity-colab-v0.2
Your branch is up to date with 'origin/experiment/fine-tune-similarity-colab-v0.2'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	Ai-Recruiter-Mini-Ai-Service/
	checkpoints/
	wandb/

nothing added to commit but untracked files present (use "git add" to track)


In [15]:
!nvidia-smi

Sun Jun  7 10:01:22 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   36C    P8              9W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

## 2. Install dependencies

In [16]:
!pip install -r requirements.txt

## 3. Validate dataset

In [17]:
!python training/validate_dataset.py --dataset-root datasets --version v0.2
!python training/check_fine_tune_readiness.py --dataset-root datasets --version v0.2

Dataset validation passed.
Pair dataset: datasets/versions/v0.2/cv_jd_pairs.jsonl
Total pairs: 2275

Fine-tuning readiness check passed.
Next action: create a separate fine-tuning experiment branch and keep generated artifacts out of git.


## 4. Smoke test

This checks that the script, dataset, and model dependencies work. Do not use this result for model comparison.

In [26]:
!python training/fine_tune_similarity.py \
  --dataset-root datasets \
  --version v0.2 \
  --base-model sentence-transformers/all-MiniLM-L6-v2 \
  --output-dir artifacts/models/debug-miniLM-v0.2 \
  --report-path artifacts/reports/debug_fine_tune_similarity_v0.2_report.json \
  --epochs 1 \
  --batch-size 2 \
  --max-train-samples 20 \
  --max-eval-samples 20

2026-06-07 10:11:49.298419: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Fine-tuning CV-JD similarity model.
Dataset version: v0.2
Base model: sentence-transformers/all-MiniLM-L6-v2
Train examples: 20
Validation examples: 20
Test examples: 20
Epochs: 1
Batch size: 2
Warmup steps: 1
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: 3
wandb: You chose "Don't visualize my results"
wandb: Using W&B in offline mode.
wandb: W&B API key is configured. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.27.0
wandb: W&B syncing is set to `offline` in this directory. Run `wandb online` or set WANDB_MODE=online to enable cloud syncing.
w

## 5. Full fine-tuning run

In [27]:
!python training/fine_tune_similarity.py \
  --dataset-root datasets \
  --version v0.2 \
  --base-model sentence-transformers/all-MiniLM-L6-v2 \
  --output-dir artifacts/models/fine-tuned-miniLM-v0.2 \
  --report-path artifacts/reports/fine_tune_similarity_v0.2_report.json \
  --epochs 1 \
  --batch-size 8

2026-06-07 10:12:29.078539: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Fine-tuning CV-JD similarity model.
Dataset version: v0.2
Base model: sentence-transformers/all-MiniLM-L6-v2
Train examples: 1600
Validation examples: 325
Test examples: 350
Epochs: 1
Batch size: 8
Warmup steps: 20
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: 3
wandb: You chose "Don't visualize my results"
wandb: Using W&B in offline mode.
wandb: W&B API key is configured. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.27.0
wandb: W&B syncing is set to `offline` in this directory. Run `wandb online` or set WANDB_MODE=online to enable cloud synci

## 6. Review report

In [28]:
import json
from pathlib import Path

report_path = Path('artifacts/reports/fine_tune_similarity_v0.2_report.json')
report = json.loads(report_path.read_text(encoding='utf-8'))
report['metrics']


{'validation': {'pair_count': 325,
  'mae': 11.6831,
  'rmse': 15.2639,
  'label_accuracy': 0.4431,
  'mean_similarity': 0.584157,
  'mean_predicted_score': 58.4277,
  'mean_target_score': 62.6892},
 'test': {'pair_count': 350,
  'mae': 11.5314,
  'rmse': 16.5403,
  'label_accuracy': 0.4857,
  'mean_similarity': 0.574352,
  'mean_predicted_score': 57.4171,
  'mean_target_score': 62.3886}}

## 7. Save artifacts to Google Drive


In [29]:
from google.colab import drive
drive.mount('/content/drive')

import shutil, pathlib
dest = pathlib.Path('/content/drive/MyDrive/ai-recruiter-artifacts')
dest.mkdir(parents=True, exist_ok=True)

shutil.copytree(
    'artifacts/models/fine-tuned-miniLM-v0.2',
    dest / 'fine-tuned-miniLM-v0.2',
    dirs_exist_ok=True
)
shutil.copy(
    'artifacts/reports/fine_tune_similarity_v0.2_report.json',
    dest / 'fine_tune_similarity_v0.2_report.json'
)
print("Saved to Drive")


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Saved to Drive


## 8. Experiment 2 - epochs=2

Epoch 1 did not beat baseline on MAE/RMSE. Trying 2 epochs.

Baseline to beat: MAE < 17.6431 | RMSE < 22.4497 | Label accuracy > 0.2822


In [30]:
!python training/fine_tune_similarity.py \
  --dataset-root datasets \
  --version v0.2 \
  --base-model sentence-transformers/all-MiniLM-L6-v2 \
  --output-dir artifacts/models/fine-tuned-miniLM-v0.2-ep2 \
  --report-path artifacts/reports/fine_tune_similarity_v0.2_ep2_report.json \
  --epochs 2 \
  --batch-size 8


2026-06-07 10:14:33.681168: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Fine-tuning CV-JD similarity model.
Dataset version: v0.2
Base model: sentence-transformers/all-MiniLM-L6-v2
Train examples: 1600
Validation examples: 325
Test examples: 350
Epochs: 2
Batch size: 8
Warmup steps: 40
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: 3
wandb: You chose "Don't visualize my results"
wandb: Using W&B in offline mode.
wandb: W&B API key is configured. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.27.0
wandb: W&B syncing is set to `offline` in this directory. Run `wandb online` or set WANDB_MODE=online to enable cloud synci

## 9. Review report - experiment 2


In [31]:
import json
from pathlib import Path

report_path = Path('artifacts/reports/fine_tune_similarity_v0.2_ep2_report.json')
report = json.loads(report_path.read_text(encoding='utf-8'))
report['metrics']


{'validation': {'pair_count': 325,
  'mae': 11.0062,
  'rmse': 14.919,
  'label_accuracy': 0.48,
  'mean_similarity': 0.612042,
  'mean_predicted_score': 61.1908,
  'mean_target_score': 62.6892},
 'test': {'pair_count': 350,
  'mae': 10.6514,
  'rmse': 15.7685,
  'label_accuracy': 0.4943,
  'mean_similarity': 0.605084,
  'mean_predicted_score': 60.5257,
  'mean_target_score': 62.3886}}

## 10. Save artifacts - Experiment 2 (epochs=2) to Google Drive

Epoch 2 performed worse than epoch 1 across all metrics. Saving artifacts for record only.

| Metric       | Baseline | Epoch 1 | Epoch 2 |
|--------------|----------|---------|---------|
| Val MAE      | 17.6431  | 19.3292 | 19.8892 |
| Val RMSE     | 22.4497  | 24.5166 | 25.033  |
| Val Acc      | 0.2822   | 0.2769  | 0.2738  |
| Test Acc     | 0.2822   | 0.3057  | 0.2971  |

Neither run beats the baseline. More epochs made results worse — sign of wrong training direction.
Do not integrate this model into service inference.


In [32]:
shutil.copytree(
    'artifacts/models/fine-tuned-miniLM-v0.2-ep2',
    dest / 'fine-tuned-miniLM-v0.2-ep2',
    dirs_exist_ok=True
)
shutil.copy(
    'artifacts/reports/fine_tune_similarity_v0.2_ep2_report.json',
    dest / 'fine_tune_similarity_v0.2_ep2_report.json'
)


PosixPath('/content/drive/MyDrive/ai-recruiter-artifacts/fine_tune_similarity_v0.2_ep2_report.json')

## 11. Experiment 3 - epochs=3

Both fixes applied (SentenceTransformerTrainer + similarity_to_score) improved results significantly.
Epoch 2 beats epoch 1 across all metrics — model is still improving. Trying epoch 3.

| Metric        | Baseline | Epoch 1  | Epoch 2  |
|---------------|----------|----------|----------|
| Val MAE       | 17.6431  | 11.6831  | 11.0062  |
| Val RMSE      | 22.4497  | 15.2639  | 14.9190  |
| Val Label acc | 0.2822   | 0.4431   | 0.4800   |
| Test MAE      | —        | 11.5314  | 10.6514  |
| Test RMSE     | —        | 16.5403  | 15.7685  |
| Test Label acc| —        | 0.4857   | 0.4943   |

Goal: check if epoch 3 continues improving.
If epoch 3 is worse than epoch 2 → epoch 2 model is optimal, stop here.


In [33]:
!python training/fine_tune_similarity.py \
  --dataset-root datasets \
  --version v0.2 \
  --base-model sentence-transformers/all-MiniLM-L6-v2 \
  --output-dir artifacts/models/fine-tuned-miniLM-v0.2-ep3 \
  --report-path artifacts/reports/fine_tune_similarity_v0.2_ep3_report.json \
  --epochs 3 \
  --batch-size 8


2026-06-07 10:21:05.826016: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Fine-tuning CV-JD similarity model.
Dataset version: v0.2
Base model: sentence-transformers/all-MiniLM-L6-v2
Train examples: 1600
Validation examples: 325
Test examples: 350
Epochs: 3
Batch size: 8
Warmup steps: 60
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: 3
wandb: You chose "Don't visualize my results"
wandb: Using W&B in offline mode.
wandb: W&B API key is configured. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.27.0
wandb: W&B syncing is set to `offline` in this directory. Run `wandb online` or set WANDB_MODE=online to enable cloud synci

## 12. Review report - experiment 3

In [34]:
import json
from pathlib import Path

report_path = Path('artifacts/reports/fine_tune_similarity_v0.2_ep3_report.json')
report = json.loads(report_path.read_text(encoding='utf-8'))
report['metrics']


{'validation': {'pair_count': 325,
  'mae': 10.88,
  'rmse': 14.8562,
  'label_accuracy': 0.48,
  'mean_similarity': 0.61405,
  'mean_predicted_score': 61.3846,
  'mean_target_score': 62.6892},
 'test': {'pair_count': 350,
  'mae': 10.82,
  'rmse': 15.7261,
  'label_accuracy': 0.4857,
  'mean_similarity': 0.607794,
  'mean_predicted_score': 60.7857,
  'mean_target_score': 62.3886}}

## 13. Save artifacts - Experiment 3 (epochs=3) to Google Drive

In [35]:
import shutil, pathlib
dest = pathlib.Path('/content/drive/MyDrive/ai-recruiter-artifacts')
dest.mkdir(parents=True, exist_ok=True)

shutil.copytree(
    'artifacts/models/fine-tuned-miniLM-v0.2-ep3',
    dest / 'fine-tuned-miniLM-v0.2-ep3',
    dirs_exist_ok=True
)
shutil.copy(
    'artifacts/reports/fine_tune_similarity_v0.2_ep3_report.json',
    dest / 'fine_tune_similarity_v0.2_ep3_report.json'
)
print("Saved to Drive")


Saved to Drive


## 14. Experiment 4 - epochs=4

Epoch 3 val improved slightly but test MAE got worse — early overfitting signal.
Improvement is diminishing. Trying epoch 4 as the final experiment.
If val MAE does not improve over epoch 3 (10.88), epoch 2 is the best model.

| Metric        | Epoch 1 | Epoch 2  | Epoch 3  |
|---------------|---------|----------|----------|
| Val MAE       | 11.6831 | 11.0062  | 10.8800  |
| Val RMSE      | 15.2639 | 14.9190  | 14.8562  |
| Val Label acc | 0.4431  | 0.4800   | 0.4800   |
| Test MAE      | 11.5314 | 10.6514  | 10.8200  |
| Test Label acc| 0.4857  | 0.4943   | 0.4857   |


In [36]:
!python training/fine_tune_similarity.py \
  --dataset-root datasets \
  --version v0.2 \
  --base-model sentence-transformers/all-MiniLM-L6-v2 \
  --output-dir artifacts/models/fine-tuned-miniLM-v0.2-ep4 \
  --report-path artifacts/reports/fine_tune_similarity_v0.2_ep4_report.json \
  --epochs 4 \
  --batch-size 8


2026-06-07 10:30:57.264111: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Fine-tuning CV-JD similarity model.
Dataset version: v0.2
Base model: sentence-transformers/all-MiniLM-L6-v2
Train examples: 1600
Validation examples: 325
Test examples: 350
Epochs: 4
Batch size: 8
Warmup steps: 80
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: 3
wandb: You chose "Don't visualize my results"
wandb: Using W&B in offline mode.
wandb: W&B API key is configured. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.27.0
wandb: W&B syncing is set to `offline` in this directory. Run `wandb online` or set WANDB_MODE=online to enable cloud synci

## 15. Review report - experiment 4


In [37]:
import json
from pathlib import Path

report_path = Path('artifacts/reports/fine_tune_similarity_v0.2_ep4_report.json')
report = json.loads(report_path.read_text(encoding='utf-8'))
report['metrics']


{'validation': {'pair_count': 325,
  'mae': 11.0031,
  'rmse': 14.8195,
  'label_accuracy': 0.4769,
  'mean_similarity': 0.613833,
  'mean_predicted_score': 61.3846,
  'mean_target_score': 62.6892},
 'test': {'pair_count': 350,
  'mae': 10.9143,
  'rmse': 15.5514,
  'label_accuracy': 0.4857,
  'mean_similarity': 0.604143,
  'mean_predicted_score': 60.4057,
  'mean_target_score': 62.3886}}

## 16. Final summary

Epoch 3 is the best model. Epoch 4 val MAE increased (10.88 → 11.0031) — overfitting confirmed.

Best model: artifacts/models/fine-tuned-miniLM-v0.2-ep3

| Metric        | Baseline | Epoch 3 (best) |
|---------------|----------|----------------|
| Val MAE       | 17.6431  | 10.8800        |
| Val RMSE      | 22.4497  | 14.8562        |
| Val Label acc | 0.2822   | 0.4800         |
| Test MAE      | —        | 10.8200        |
| Test Label acc| —        | 0.4857         |

Fine-tuning experiment complete. Proceed to integration branch.


## 17. Save artifacts - Experiment 4 (epochs=4) to Google Drive

Epoch 4 is worse than epoch 3 — saving for record only.


In [39]:
import shutil, pathlib
dest = pathlib.Path('/content/drive/MyDrive/ai-recruiter-artifacts')

shutil.copytree(
    'artifacts/models/fine-tuned-miniLM-v0.2-ep4',
    dest / 'fine-tuned-miniLM-v0.2-ep4',
    dirs_exist_ok=True
)
shutil.copy(
    'artifacts/reports/fine_tune_similarity_v0.2_ep4_report.json',
    dest / 'fine_tune_similarity_v0.2_ep4_report.json'
)
print("Saved to Drive")


Saved to Drive
